### PR TITLE
CONTRIBUTING: Add note about how to commit new maintainers-list.nix entry

### PR DESCRIPTION
**Pull request by matthiasbeyer**
In #247972 I found that the contributing file does not say how to add a maintainer, only the manual does (as @ius said in above PR).

Maybe we can add a note here.

### CODE REVIEW

**Review by tjni — APPROVED** [review]

**Comment by Janik-Haag** [comment]
this probably relates to #245243

**Review by ius — DISMISSED** [review]
Thanks, beat me to it.

Two remarks:

1. `<your github handle>`: just `handle` I think, see [this remark](https://github.com/NixOS/nixpkgs/blob/master/maintainers/maintainer-list.nix#L21) - it doesn't have to be a GitHub username (not all GH names are valid attribute anyway, I think?)

2. Related: I wonder whether it should be explicit about submitting it as a separate commit. I think it makes sense and I see many reviewers request this, but on the other hand most committers/the GitHub default end up creating a merge commit, thus hiding said commit?

**Comment by matthiasbeyer** [comment]
Fixed what you suggested!

I will squash this if someone requests it.

**Review by infinisil — CHANGES_REQUESTED** [review]

**Comment by infinisil** [comment]
> this probably relates to #245243

Thanks for spreading awareness, I'd love some reviews for that once it's not a draft anymore btw, feel free to subscribe to it!

That PR is still a bit too much in flux though, I don't think it should be blocking this PR here. I can rebase my PR once this one is merged (yeah it's a bit annoying to rebase when things have been moved, but I can manage :wink:)

**Comment by Lord-Valen** [comment]
> 2. Related: I wonder whether it should be explicit about submitting it as a separate commit. I think it makes sense and I see many reviewers request this, but on the other hand most committers/the GitHub default end up creating a merge commit, thus hiding said commit?

I think it should be. Also, merge commits don't hide other commits if what you mean by "hide" is something that implies that they somehow aren't in the branch. A merge commit adds a new commit with two parent commits (the latest commit of each branch) that resolves any conflicts. Both parent commits are included in the receiving branch. You might think that merge commits hide other commits since the dates of those commits don't change and the merge commit diff shows the same changes. A merge commit can occur any amount of time after its parents, which can result in a large distance between them. With the velocity of Nixpkgs, that usually means the merge commit and one of the parents have quite a few commits between them. Here's an example: https://github.com/NixOS/nixpkgs/commit/195975b1387b69ac5566338819298d7abc595ac3 merges  https://github.com/NixOS/nixpkgs/commit/abc3cc0ac6f44ebbd997464c2e1460c9001d7743 into master on https://github.com/NixOS/nixpkgs/commit/e04428a3a8e4e672314011c07d56c18b62e34e89.
Right now there are 22 commits on master in the time between https://github.com/NixOS/nixpkgs/commit/195975b1387b69ac5566338819298d7abc595ac3 and https://github.com/NixOS/nixpkgs/commit/abc3cc0ac6f44ebbd997464c2e1460c9001d7743.

If you want to see an example of a merge commit for a PR with an additional `maintainers:` commit, you could explore the parents of https://github.com/NixOS/nixpkgs/commit/36ea4a4b6ab950c3ea1ea9d7fe036f35bd85d2cb.

File: CONTRIBUTING.md
Comment (Lord-Valen) [reply CONTRIBUTING.md:64]: Yes, I would put the commit message format in as an example and change the template accordingly under "Submitting Changes" and also try to make it more obvious that changes to `maintainer-list.nix` should occur in a separate commit.

I think it might be reasonable to also normalize `doc/<page>:` and `lib/<module>:` commit messages or simply `doc:` and `lib:` for more general changes, but that isn't necessary here.

Comment (infinisil) [reply CONTRIBUTING.md:64]: Ah yeah then I guess this fits very well into the scope of the contributing documentation cleanup in #245243, because I create separate contributing documents for each part of Nixpkgs, such as `pkgs/README.md`, `doc/README.md`, `maintainers/README.md`, etc. And this would fit very well into `maintainers/README.md`.

Comment (matthiasbeyer) [reply CONTRIBUTING.md]: I added the change requested by @infinisil as a commit to this PR, which I can squash if requested of course.

File: CONTRIBUTING.md
Comment (AndersonTorres) [inline CONTRIBUTING.md:40]:
```suggestion
  In case of adding yourself as maintainer, do it in a separated commit (no need for a Pull Request), with message `maintainers: add <handle>`. See also ([the section of manual about meta.maintainers](https://nixos.org/manual/nixpkgs/stable/#var-meta-maintainers).
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: CONTRIBUTING.md
Comment (Lord-Valen) [reply CONTRIBUTING.md:40]: It isn't even necessarily adding yourself. It's adding anyone to the maintainer list.

Comment (AndersonTorres) [reply CONTRIBUTING.md:40]: This is the most common scenario: a newcomer adding a package.
But point taken, I will update this.

File: CONTRIBUTING.md
Comment (fricklerhandwerk) [inline CONTRIBUTING.md:40]:
```suggestion
  When adding yourself as maintainer, in the same pull request make a separate commit with the message `maintainers: add <handle>`.
  See [Nixpkgs Maintainers](../maintainers/README.md) for details.
```

Given the information will be expanded in https://github.com/NixOS/nixpkgs/pull/250344, we should rather link to that file. We can still add information on NixOS contributions later.

**Comment by VuiMuich** [comment]
On my PR adding myself to `maintainers` I got advised to reorder the commits, so it is before the package commit. Might be worth adding a side note for this being preferred. 
Idk if this technically makes any CI checks easier (fail earlier in eval maybe?), but logically it makes sense at least.

File: CONTRIBUTING.md
Comment (fricklerhandwerk) [inline CONTRIBUTING.md:40]:
```suggestion
  When adding yourself as maintainer, in the same pull request make a separate commit with the message `maintainers: add <handle>`.
Add the commit before those making changes to the package or module.
```

**Review by fricklerhandwerk — APPROVED** [review]
@matthiasbeyer please rebase.

File: CONTRIBUTING.md
Comment (VuiMuich) [reply CONTRIBUTING.md:40]: Is it intentional, or is the indentation off?

**Review by infinisil — CHANGES_REQUESTED** [review]
This is in the wrong place now, should be in `maintainers/README.md`

**Comment by fricklerhandwerk** [comment]
@Infinisil: this section is about commit formatting. The organizational details are in the other file. I think it makes sense to have it here, and also use the opportunity to direct people to the new extended page.

**Comment by infinisil** [comment]
True there are commit conventions in `CONTRIBUTING.md`, but they're [further down](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions). I'd be fine with it there, but I do think it should be in `maintainers/README.md` in the future (and same for package commit conventions, which should be in `pkgs/README.md`). But the PR here currently adds it to the PR template section, which is definitely wrong.

**Comment by VuiMuich** [comment]
I agree, that it would be a better fit in the "How to create pull requests" section, probably as part of bullte 4.

Imo it would not hurt to double this small information in both the `CONTRIBUTING.md` and `maintainers/README.md` as it is easy to be overlooked.

File: CONTRIBUTING.md
Comment (fricklerhandwerk) [reply CONTRIBUTING.md]: Ech, no, that was a typo.

**Comment by infinisil** [comment]
@VuiMuich I'd rather not have it be duplicated. It makes it harder to know which part should be updated, or if you don't know about both parts they could get outdated. I think it's better to improve the linking between the sections.

> I agree, that it would be a better fit in the "How to create pull requests" section, probably as part of bullet 4.

Step 4 links to the commit conventions [here](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions), where there's already information about the commit message formatting.

File: CONTRIBUTING.md
Comment (cafkafk) [inline CONTRIBUTING.md:158]:
```suggestion
When adding yourself as maintainer, in the same pull request, make a separate commit with the message `maintainers: add <handle>`.
```

**Review by cafkafk — APPROVED** [review]
Comma nitpick but lgtm

**Comment by VuiMuich** [comment]
I can follow that "harder to update" argument very well and as I have stumbled on a bunch of such discrepancies between different docs in the bigger nix eco system. With this in mind, I agree now, to better avoid this duplication.

Also putting it as a bullet in the "commit conventions" totally makes sense.

**Comment by fricklerhandwerk** [comment]
@matthiasbeyer can you follow up with the changes? Since there seems to be a lot of energy behind this right now, if you lack time the docs team can take this over.

**Comment by matthiasbeyer** [comment]
I was traveling today, I can have a look tomorrow!

**Comment by matthiasbeyer** [comment]
Squashed the comma commit from @cafkafk and added a new one moving the text block to "commit conventions" as per request.

I can squash these commits as well, if you like!

**Comment by cafkafk** [comment]
It probably makes most sense to have just a single commit for this, instead of one adding it and another moving it.

**Comment by asymmetric** [comment]
Not sure if here or in another PR, but it would be good to also document how to add oneself to `meta.maintainers`, meaning just what you did in this PR:

- how the commit should be formatted
- the placement (together with a commit making a change to a package, or separately?)

Same for removal probably.

**Comment by infinisil** [comment]
@asymmetric It's mostly there in `maintainers/README.md`, though it could use a bit of work

**Review by fricklerhandwerk — APPROVED** [review]
Let's merge this first and follow up afterwards.

**Review by AndersonTorres — APPROVED** [review]

**Review by infinisil — APPROVED** [review]

**Pull request merged**
